internal/platform/state/contract/evolution/contract.go

```diff
@@ -24,29 +25,33 @@ func NewService(tx storage.Tx) *service {
 }
 
 func (s *service) StoreMintedWithExternalURIEvents(contract string, events []model.MintedWithExternalURI) error {
-	var buf bytes.Buffer
-	encoder := gob.NewEncoder(&buf)
-
-	if err := encoder.Encode(events); err != nil {
-		return err
+	for _, event := range events {
+		var buf bytes.Buffer
+		encoder := gob.NewEncoder(&buf)
+		if err := encoder.Encode(event); err != nil {
+			return err
+		}
+		if errSet := s.tx.Set([]byte(eventsPrefix+strings.ToLower(contract)+"_"+fmt.Sprint(event.BlockNumber)+"_"+fmt.Sprint(event.TxIndex)), buf.Bytes()); errSet != nil {
+			return errSet
+		}
 	}
-
-	return s.tx.Set([]byte(eventsPrefix+strings.ToLower(contract)), buf.Bytes())
+	return nil
 }
 
 func (s *service) GetMintedWithExternalURIEvents(contract string) ([]model.MintedWithExternalURI, error) {
-	value, err := s.tx.Get([]byte(eventsPrefix + strings.ToLower(contract)))
-	if err != nil {
-		return nil, err
-	}
-	if value == nil {
-		return nil, nil
+	events := s.tx.GetValuesWithPrefix([]byte(eventsPrefix + strings.ToLower(contract) + "_"))
```

Comment:
this line loads all the events. Probably it does not affect transaction size because it is just read, but at some point this is also going to be an issue (slow load) if the number of events becomes significantly big.  We should think about how to optimize this, also. Not sure if is it easy to fix. if not I would create jira task and leave it for the next sprint.

Reply:
jira task created: https://freeverse.atlassian.net/browse/UN-137